### ISSUE TITLE
Edge Formats

### ISSUE THREAD

**Issue by yishaiSilver** [cgraber/cvpr_dNRI#8]
Hello,

I've noticed that the edges generated by the NRI paper are in a symmetric matrix format, whereas the edges expected by dNRI are in a vector format. Likewise, it seems that -- for n particles -- the vector is of length n * (n-1), which makes sense since each particle's relation to itself is a given. That being the case, if I flatten the matrix (excluding the diagonal values) and pass that in as the vector, the edge predictions are always about a 50-50 guess. Any help would be greatly appreciated.

Thanks.

**Comment by cgraber** [comment]
Hi Yishai,

We construct the edges within the encoder [in these lines](https://github.com/cgraber/cvpr_dNRI/blob/master/dnri/models/dnri.py#L275-L278). Specifically, we start with a full ones matrix and subtract out the diagonal, and then essentially "flatten" to get a vector of pairs.

As a quick check, I ran inference using a pre-trained model on the basketball data, and I looked at the predicted logits [at this line](https://github.com/cgraber/cvpr_dNRI/blob/master/dnri/models/dnri.py#L119). The shape of the logits is [batch_size, input_len-1, num_pairs, num_edge_types], where num_pairs = 20 in my case (since there are 5 players in this data). The predictions I got for one time step for one batch entry were:
          [ 4.6055, -4.5495],
          [ 4.3164, -4.3480],
          [ 2.4512, -3.1706],
          ...,
          [ 1.4313, -2.0851],
          [-0.1860,  0.8247],
          [-3.5411,  2.9512]

so I can confirm that I'm not getting a uniform distribution.

So that I can help further: which data are use using? Is this one of the datasets we used in our paper, or is it some custom dataset? Did you train using our training script, or are you doing something else?

**Comment by yishaiSilver** [comment]
Hello again, 

Okay, then I was understanding the edge shape/format correctly, so that's good to hear.

As for my dataset, I'm using a dataset generated with NRI: https://github.com/ethanfetaya/NRI/blob/master/data/generate_dataset.py. To be clear, it has static edges, which I assumed would help me debug my issue with dNRI.

And yes, I'm using your training script. The only modification I've made is adding the eval_edges function from the ..._experiment.py file. I did this to see how the edge predictions were changing over time. 

And then, finally, my results are particularly weird. They're almost always predicting one edge type:

        [[[ -4.4010,   4.6667],
          [  0.6088,  -0.3821],
          [ -5.5756,   7.0464],
          ...,
          [  3.7157,  -4.7355],
          [  5.5048,  -7.0182],
          [  3.2469,  -4.1171]],

         [[ -7.4909,   8.2328],
          [ -7.2132,   8.5512],
          [ -8.6809,  10.6069],
          ...,
          [ -7.7252,   8.7680],
          [ -7.4723,   8.3750],
          [ -8.6295,  10.0191]],

         [[ -9.4068,  10.4179],
          [-10.0142,  11.5074],
          [ -9.9098,  11.9369],
          ...,
          [ -9.9795,  11.1857],
          [ -9.7064,  10.7637],
          [-10.1917,  11.6953]],

         ...,

         [[-11.0803,  12.4869],
          [-10.0659,  11.9223],
          [ -8.6793,  10.4833],
          ...,
          [-11.5727,  13.1095],
          [-10.0299,  12.0409],
          [ -9.2140,  11.0667]],

         [[-11.1848,  12.6869],
          [-10.1493,  11.9822],
          [ -8.5420,  10.3441],
          ...,
          [-11.6719,  13.2544],
          [-10.2817,  12.2809],
          [ -9.0036,  10.8821]],

         [[-11.1345,  12.7730],
          [-10.2828,  12.0995],
          [ -8.5565,  10.3753],
          ...,
          [-11.6273,  13.2500],
          [-10.5280,  12.5249],
          [ -8.9474,  10.8262]]],

Val Edge results:
	F1:  0.6643718522801465
	All predicted edge accuracy:  0.5000667708333328
	First Edge Acc:  0.01052189604587582
	Second Edge Acc:  0.9894158668266325

Please note that the edges are predicted with a 50-50 chance and then diverge to about a 1-99 chance as training goes on. Regardless, the average accuracy remains 50%.

**Comment by cgraber** [comment]
Some more information would be useful:

- Which simulation setting are you using? How many variables? How many variables are connected?
- What training flags are you using?

If I'm understanding the data correctly, the "first edge type" represents no connection and the "second edge type" represents a connection (this is how my code generally interprets edge types, anyways). So these kinds of stats seem to indicate that the model is predicting an edge (i.e. "the second edge type") between all pairs of variables. How well does the static NRI model do on this task?

**Comment by yishaiSilver** [comment]
I used the default parameters to generate the dataset: spring, 5 particles, and a random number of connected particles for each simulation.

My training flags are here:

> WORKING_DIR="${BASE_RESULTS_DIR}/dnri/seed_${SEED}/"
ENCODER_ARGS="--encoder_hidden 256 --encoder_mlp_num_layers 3 --encoder_mlp_hidden 128 --encoder_rnn_hidden 64"
DECODER_ARGS="--decoder_hidden 256"
HIDDEN_ARGS="--rnn_hidden 64"
PRIOR_ARGS="--use_learned_prior --prior_num_layers 3 --prior_hidden_size 128"
MODEL_ARGS="--model_type dnri --graph_type dynamic --skip_first --num_edge_types 2 $ENCODER_ARGS $DECODER_ARGS $HIDDEN_ARGS $PRIOR_ARGS --seed ${SEED}"
TRAINING_ARGS='--batch_size 256 --lr 5e-4 --use_adam --num_epochs 100 --normalize_kl --normalize_nll --tune_on_nll --val_teacher_forcing --teacher_forcing_steps -1'
mkdir -p $WORKING_DIR
CUDA_VISIBLE_DEVICES=$GPU python -u dnri/experiments/spring_experiment.py --gpu --mode train --data_path $DATA_PATH --working_dir $WORKING_DIR $MODEL_ARGS $TRAINING_ARGS | tee "${WORKING_DIR}results.txt"
CUDA_VISIBLE_DEVICES=$GPU python -u dnri/experiments/spring_experiment.py --gpu --mode eval --load_best_model --data_path $DATA_PATH --working_dir $WORKING_DIR $MODEL_ARGS $TRAINING_ARGS | tee "${WORKING_DIR}eval_results.txt"

As for NRI, it performed well and mirrored their paper's results:
> Epoch: 0177 nll_train: 2.2486116141 kl_train: -0.0030197995 mse_train: 0.0000011712 acc_train: 0.9981461797 nll_val: 0.0790408968 kl_val: -0.0033634473 mse_val: 0.0000000412 acc_val: 0.9976958070 time: 1009.9070s

**Comment by cgraber** [comment]
How do the NLL/MSE for dNRI compare against those for static NRI? 

I'll also mention that, if the interaction graph is expected to be static, then I'd expect static NRI to perform better - this model is able to average the interactions across time to come up with its interaction prediction, while dNRI has to make a separate analysis at every point in time.
